File: libraries/SQLStatement.go
package main

import "container/list"
// import "fmt"
import "ocl"
// import "strings"
// import "math"
import "reflect"
import "ocliterator"
import "ocldate"

import "database/sql"


type SQLStatement struct {
  text string
  actualStatement *sql.Stmt
  databaseConnection *OclDatasource
  parameterMap map[int]interface{}
  parameters []interface{}
  maxfield int
  resultSet *ocliterator.OclIterator
}

var SQLStatement_instances = list.New()


func createSQLStatement() *SQLStatement {
  var res *SQLStatement
  res = &SQLStatement{}
  SQLStatement_instances.PushBack(res)
  return res
}

func NewSQLStatement() *SQLStatement {
  res := createSQLStatement()
  res.actualStatement = nil
  res.databaseConnection = nil
  res.parameterMap = make(map[int]interface{})
  res.resultSet = nil
  res.maxfield = 0
  return res
}

var TYPESQLStatement = reflect.TypeOf(&SQLStatement{})

func (self *SQLStatement) SetStatement(s *sql.Stmt) { 
   self.actualStatement = s 
}

func (self *SQLStatement) SetConnection(c *OclDatasource) { 
   self.databaseConnection = c 
}

func (self *SQLStatement) Close() {
  if self.actualStatement != nil { 
    self.actualStatement.Close()
    self.actualStatement = nil
  } 
  self.parameterMap = make(map[int]interface{})
  self.parameters = nil
  self.resultSet = nil
  self.maxfield = 0
}

func (self *SQLStatement) CloseOnCompletion() {

}

func (self *SQLStatement) SetString(field int,  value string) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetInt(field int,  value int) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetByte(field int,  value int) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetShort(field int,  value int) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetBoolean(field int,  value bool) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetLong(field int,  value int64) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetDouble(field int,  value float64) {
  self.parameterMap[field] = value
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetTimestamp(field int,  value *ocldate.OclDate) {
  self.parameterMap[field] = value.GetTime()
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) SetNull(field int,  value interface{}) {
  self.parameterMap[field] = nil
  if field > self.maxfield { 
    self.maxfield = field
  } 
}

func (self *SQLStatement) parameterMapToFields() {
    self.parameters = make([]interface{}, self.maxfield)
    for x := 1; x <= self.maxfield; x++ {
      val := self.parameterMap[x] 
      self.parameters[x-1] = val
    } 
}

func (self *SQLStatement) ExecuteUpdate() {
  if self.actualStatement != nil { 
    self.actualStatement.Exec()
  } 
}

func (self *SQLStatement) query_String_Array(parvals []interface{}) *ocliterator.OclIterator {
  var result *ocliterator.OclIterator = nil
  if self.actualStatement != nil { 
    records := list.New()
    rows, _ := self.actualStatement.Query(parvals...)
    
    defer rows.Close()

    var cols []string

    for rows.Next() { 
      cols, _ = rows.Columns()
      m := len(cols) 
      row := make(map[string]interface{})
      values := make([]interface{}, m)
      for i := 0; i < m; i++ { 
        var xi interface{}
        values[i] = &xi
      }
      rows.Scan(values...)
      for i := 0; i < m; i++ { 
        ptr := values[i].(*interface{})
        row[cols[i]] = *ptr
      } 
      records.PushBack(row)
    } 

    result = ocliterator.NewOclIterator_Sequence(records)
    n := len(cols)
    colNames := make([]interface{}, n)
    for i := 0; i < n; i++ { 
      colNames[i] = cols[i]
    } 
    result.SetColumnNames(ocl.SequenceRange(colNames))
  } 

  return result
}

func (self *SQLStatement) ExecuteQuery(stat ...string) *ocliterator.OclIterator {
  var result *ocliterator.OclIterator = nil
  if len(stat) == 0 { 
    self.parameterMapToFields()
    self.resultSet = self.query_String_Array(self.parameters)
    return self.resultSet
  } else { 
    ss := stat[0]
    if self.databaseConnection != nil { 
      self.resultSet = self.databaseConnection.Query_String(ss)
      return self.resultSet
    } 
  } // execute on the databaseConnection
  return result
}

func (self *SQLStatement) update_String_Array(parvals []interface{}) {
  if self.actualStatement != nil { 
    self.actualStatement.Exec(parvals...)
  }
}

func (self *SQLStatement) Execute(stat ...string) {
  if len(stat) == 0 { 
    self.parameterMapToFields()
    self.update_String_Array(self.parameters)
  } else { 
    ss := stat[0]
    if self.databaseConnection != nil { 
      self.databaseConnection.ExecSQL(ss)
    } 
  } // execute on the databaseConnection
}

func (self *SQLStatement) Cancel() {
}

func (self *SQLStatement) GetConnection() *OclDatasource {
  return self.databaseConnection
}

func (self *SQLStatement) GetResultSet() *ocliterator.OclIterator {
  return self.resultSet
}

File: libraries/SQLStatement.swift
import Foundation
import Darwin
import SQLite3

class SQLStatement
{ private static var instance : SQLStatement? = nil
  var text : String = ""
  var statement: OpaquePointer?
  var dbPointer : OpaquePointer?
  var database : OclDatasource?
  var resultSet : OclIterator = OclIterator.defaultInstance() 
    
  init() { }

  init(copyFrom: SQLStatement) {
    self.text = copyFrom.text
    self.statement = copyFrom.statement
    self.dbPointer = copyFrom.dbPointer
    self.resultSet = copyFrom.resultSet
    self.database = copyFrom.database
  }

  func copy() -> SQLStatement
  { let res : SQLStatement = SQLStatement(copyFrom: self)
    addSQLStatement(instance: res)
    return res
  }

  static func defaultInstanceSQLStatement() -> SQLStatement
  { if (instance == nil)
    { instance = createSQLStatement() }
    return instance!
  }

  deinit
  { killSQLStatement(obj: self) }

  func close() -> Void
  {
    if (statement != nil)
    { sqlite3_finalize(statement!) } 
    if (dbPointer != nil)
    { sqlite3_close(dbPointer!) } 
  }


  func closeOnCompletion() -> Void
  {

  }

/* 
func insert() {
  var insertStatement: OpaquePointer?
  // 1
  if sqlite3_prepare_v2(db, insertStatementString, -1, &insertStatement, nil) == 
      SQLITE_OK {
    let id: Int32 = 1
    let name: NSString = "Ray"
    // 2
    sqlite3_bind_int(insertStatement, 1, id)
    // 3
    sqlite3_bind_text(insertStatement, 2, name.utf8String, -1, nil)
    // 4
    if sqlite3_step(insertStatement) == SQLITE_DONE {
      print("\nSuccessfully inserted row.")
    } else {
      print("\nCould not insert row.")
    }
  } else {
    print("\nINSERT statement is not prepared.")
  }
  // 5
  sqlite3_finalize(insertStatement)
}
 */
    
    func setParameters(pars : [Any])
    { for (i,x) in pars.enumerated() {
        setObject(field: i+1, value: x)
      }
    }
    
    func setObject(field : Int, value : Any) -> Void
    {
      if (statement != nil)
      { sqlite3_bind_value(statement!, Int32(field), value as? OpaquePointer) }
    }


  func setString(field : Int, value : String) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_text(statement!, Int32(field), value, -1, nil) }
  }


  func setInt(field : Int, value : Int) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_int(statement!, Int32(field), Int32(value)) }
  }


  func setByte(field : Int, value : Int) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_int(statement!, Int32(field), Int32(value)) }
  }


  func setShort(field : Int, value : Int) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_int(statement!, Int32(field), Int32(value)) }
  }


  func setBoolean(field : Int, value : Bool) -> Void
  { guard statement != nil 
    else 
    { return }

    if value == true 
    { sqlite3_bind_int(statement, Int32(field), 1) }
    else 
    { sqlite3_bind_int(statement, Int32(field), 0) }
  }


  func setLong(field : Int, value : Int64) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_int64(statement!, Int32(field), value) }
  }


  func setDouble(field : Int, value : Double) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_double(statement!, Int32(field), value) }
  }


  func setTimestamp(field : Int, value : OclDate) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_int64(statement!, Int32(field), value.getTime()) }
  }


  func setNull(field : Int, value : Any) -> Void
  {
    if (statement != nil) 
    { sqlite3_bind_null(statement!, Int32(field)) }
  }


  func executeUpdate() throws -> Void
  {
    guard statement != nil
    else 
    { try database!.execSQL(stat: text)
      return
    }

    sqlite3_step(statement!)    
  }


  func executeQuery(stat : String) throws -> OclIterator
  {
    guard dbPointer != nil
    else 
    { return OclIterator.newOclIterator_Sequence(sq: []) }

    resultSet = try database!.query_String(stat: stat)
    return resultSet
  }


  func executeQuery() throws -> OclIterator
  { let res = try executeQuery(stat: text)
    return res
  }


  func execute(stat : String) throws -> Void
  {
    guard dbPointer != nil
    else 
    { return }

    try database!.execSQL(stat: stat)
  }


  func execute() throws -> Void
  {
    try executeUpdate()
  }


  func cancel() -> Void
  {
    if statement != nil 
    { sqlite3_reset(statement!) }
  }


  func getConnection() -> OclDatasource
  {
    return database!
  }


  func getResultSet() -> OclIterator
  {
    return resultSet
  }

}




var SQLStatement_allInstances : [SQLStatement] = [SQLStatement]()

func createSQLStatement() -> SQLStatement
{ let result : SQLStatement = SQLStatement()
  SQLStatement_allInstances.append(result)
  return result
}

func addSQLStatement(instance : SQLStatement)
{ SQLStatement_allInstances.append(instance) }

func killSQLStatement(obj: SQLStatement)
{ SQLStatement_allInstances = SQLStatement_allInstances.filter{ $0 !== obj } }


